plugin.js

```diff
@@ -0,0 +1,112 @@
+'use strict';
+
+const PLUGIN_ID_KEY = '___shelljs_plugin_id___';
+const DEFAULT_ERROR_CODE = 1;
+
+function setPluginID(plugin, id) {
+  Object.defineProperty(plugin, PLUGIN_ID_KEY, {
+    value: id
+  });
+  return id;
+}
+
+const getPluginID = plugin => plugin[PLUGIN_ID_KEY];
+
+module.exports = function setupShellPlugins($) {
+  const used_plugins = [];
+  const plugins = [{ // $
+    id: 0,
+    name: '',
+    fullname: '',
+    func: $,
+    target: null
+  }];
+  setPluginID($, 0);
+
+  const state = {
+    error: null,
+    errorCode: 0,
+    currentCmd: 'shell.js',
+    tempDir: null,
+  };
+  $.config = {
+    silent: false,
+    fatal: false,
+    verbose: false,
+    noglob: false,
+    globOptions: false,
+    maxdepth: false,
+  };
+
+
+  const utils = {
```

Comment:
Most of what was in common.expand will now go here, which is an object of utilities passed into every plugin. It's referred to in documentation and comments as `plugin.utils`.